Implement 'Save As'

vince-fugnitto opened the pull request:
Fixes #799

This PR consists of:

1. Implemented the `Save As` functionality and command in order
to save a source URI to a target destination. The `URI` in question
can either be a single file URI or a folder URI (save all files under the
directory to a target).

2. Updated the `SaveDialogProps` to be able to pass an `inputValue` to
the html text input. This is useful when providing a programmatic name
to the input if available. In the instance of a `Save` dialog it is used
to show which `URI` currently has the context.

3. Fix bug when rendering the `SaveDialog` tree icons.
Incorrectly we were using the passed `URI` to determine the correct root icon
instead of using the root `FileStat` which provides the appropriate icon.

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

akosyakov commented on packages/filesystem/src/browser/file-dialog/file-dialog.ts:
```diff
@@ -104,6 +104,11 @@ export class SaveFileDialogProps extends FileDialogProps {
+    /**
+     * A human-readable value for the input.
+     */
+    inputValue?: string;
```
Should not it be used in electron implementation somehow?

vince-fugnitto replied:
I need to see why Electron does not display the menu item but the keybinding works and triggers the dialog.
I would like to be able to pass the input value to Electron as well.

akosyakov commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -401,6 +422,43 @@ export class WorkspaceCommandContribution implements CommandContribution {
+        }
+    }
+
+    private async confirmOverwrite(uri: URI): Promise<boolean> {
```
How about implementing it in `WorkspaceFrontendContribution` then you don't need to copy code and create new keybinding contributions.

akosyakov commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -125,6 +125,11 @@ export namespace WorkspaceCommands {
+    export const SAVE_AS: Command = {
+        id: 'file.saveAs',
+        category: 'File',
+        label: 'Save As'
```
`Save As...`

akosyakov commented:
Top-level folder icon is strange:
<img width="598" alt="Screen Shot 2019-04-01 at 17 59 43" src="https://user-images.githubusercontent.com/3082655/55341996-05335580-54a8-11e9-8fca-246cd46858ba.png">

We need to test it with Electron.

vince-fugnitto commented on packages/filesystem/src/browser/file-dialog/file-dialog-service.ts:
```diff
@@ -74,7 +74,7 @@ export class DefaultFileDialogService {
     protected async getRootNode(folderToOpen?: FileStat): Promise<DirNode | undefined> {
         const folder = folderToOpen || await this.fileSystem.getCurrentUserHome();
```
@elaihau I wonder why this is the case, we already pass the folder as reference when triggering the file dialog why do we need to get the parent after? Maybe you can clarify or help me understand? Perhaps I am calling the `showSaveDialog` with the incorrect `FileStat`.

Without the change the following was displayed https://github.com/theia-ide/theia/pull/4771#issuecomment-478640031 but after we are able to have the context of what folder we are currently in and have the proper tree.

<img width="678" alt="Screen Shot 2019-04-01 at 7 59 47 PM" src="https://user-images.githubusercontent.com/40359487/55366903-bb9f3680-54b8-11e9-96b1-1b79619abe0d.png">

vince-fugnitto commented on packages/filesystem/src/browser/file-dialog/file-dialog-service.ts:
```diff
@@ -74,7 +74,7 @@ export class DefaultFileDialogService {
     protected async getRootNode(folderToOpen?: FileStat): Promise<DirNode | undefined> {
         const folder = folderToOpen || await this.fileSystem.getCurrentUserHome();
         if (folder) {
-            const rootUri = new URI(folder.uri).parent;
```
@elaihau I wonder why this is the case, we already pass the folder as reference when triggering the file dialog why do we need to get the parent after? Maybe you can clarify or help me understand? Perhaps I am calling the `showSaveDialog` with the incorrect `FileStat`.

Without the change the following was displayed https://github.com/theia-ide/theia/pull/4771#issuecomment-478640031 but after we are able to have the context of what folder we are currently in and have the proper tree.

<img width="678" alt="Screen Shot 2019-04-01 at 7 59 47 PM" src="https://user-images.githubusercontent.com/40359487/55366903-bb9f3680-54b8-11e9-96b1-1b79619abe0d.png">

akosyakov replied:
@vince-fugnitto Could you revert this change and try to pass a file URI instead of parent folder? I think it was done on purpose to allow opening siblings folders.

vince-fugnitto commented:
@akosyakov I found the bug that was leading to the incorrect root icon being displayed in the tree. 
Instead of using the correct `rootUri` stat, we were using the one passed by the function. This meant that if a file `URI` was being used, this is the icon we would be displaying as the root. Instead, I determine the icon based on the true parent directly.

https://github.com/theia-ide/theia/pull/4771/files#diff-223b9769cca26ce645f9820a5d125006R81

vince-fugnitto commented:
@akosyakov should we cover the electron dialog in this PR or as a separate PR? 
I'll have to investigate and maybe talk with @kittaakos on how it can be implemented.

akosyakov commented:
I've tested that in Electron, it does use native dialogs at least. Although not like in VS Code.

<img width="1026" alt="Screen Shot 2019-04-03 at 15 15 33" src="https://user-images.githubusercontent.com/3082655/55481814-751a1b00-5623-11e9-9fe2-b27eb4639f42.png">
<img width="1286" alt="Screen Shot 2019-04-03 at 15 14 37" src="https://user-images.githubusercontent.com/3082655/55481815-751a1b00-5623-11e9-9381-a55ae2b24dd6.png">


I also noticed that `Save Workspace As...` just does not work in Electron if i open Theia repo.

vince-fugnitto commented:
> browser part works good, electron works somehow, please at least file an issue to make it good

@akosyakov sure I can do that, we would also need to adjust the `Save Workspace As...` dialog as well. I assume the `Open` dialogs are properly implemented in Theia but the `Save` ones need some improvement.

vince-fugnitto commented:
@akosyakov btw when clicking the little arrow the same dialog can be seen as in VSCode.
I assume we don't trigger it by default like VSCode.

<div align='center'>

<img width="1085" alt="Screen Shot 2019-04-03 at 9 40 19 AM" src="https://user-images.githubusercontent.com/40359487/55483473-a5969080-55f4-11e9-8f07-6707136859a9.png">


</div>

akosyakov commented:
@vince-fugnitto cool, but input value should be different as well

kittaakos commented:
> electron works somehow, please at least file an issue to make it good

Why do we merge half-baked state? There are downstream projects using Theia's electron application and if they want to update from `next` they must pull this half ready implementation too.

BTW, the Save As in VS Code looks like this for me:

![Screen Shot 2019-04-03 at 16 23 59](https://user-images.githubusercontent.com/1405703/55486679-0b067380-562d-11e9-9224-49512b3d9e34.jpg)

vince-fugnitto commented:
@akosyakov I was able to pass the input value to electron for the file name 😄 

<div align='center'>

<img width="953" alt="Screen Shot 2019-04-03 at 10 24 16 AM" src="https://user-images.githubusercontent.com/40359487/55486708-cb269880-55fa-11e9-90c4-60b75f8a43c5.png">


</div>

vince-fugnitto commented:
> > electron works somehow, please at least file an issue to make it good
> 
> Why do we merge half-baked state? There are downstream projects using Theia's electron application and if they want to update from `next` they must pull this half ready implementation too.
> 
> BTW, the Save As in VS Code looks like this for me:
> 
> ![Screen Shot 2019-04-03 at 16 23 59](https://user-images.githubusercontent.com/1405703/55486679-0b067380-562d-11e9-9224-49512b3d9e34.jpg)

I didn't want to merge it yet either @kittaakos, that's why I spent some time trying to improve it first :)

akosyakov commented:
@kittaakos Could you finish Electron review please?

kittaakos commented:
I cannot see `Save As...` in the `File` menu. It is visible from the Command Palette. Is it a bug or a feature?

```
Akoss-MacBook-Pro:theia akos.kitta$ git rev-parse --short HEAD
aee39e5c
Akoss-MacBook-Pro:theia akos.kitta$ git status
On branch vf/save-as
Your branch is up to date with 'origin/vf/save-as'.

nothing to commit, working tree clean
Akoss-MacBook-Pro:theia akos.kitta$
```

![Screen Shot 2019-04-04 at 10 00 01](https://user-images.githubusercontent.com/1405703/55539239-7cdbcd00-56c0-11e9-9ab7-b48bda4e8f52.jpg)
![Screen Shot 2019-04-04 at 09 58 22](https://user-images.githubusercontent.com/1405703/55539240-7d746380-56c0-11e9-9321-5c35ff504255.jpg)

kittaakos commented:
If I accept the proposed file name (`inputValue`), nothing happens however, VS Code tells me the following:
![Screen Shot 2019-04-04 at 10 03 11](https://user-images.githubusercontent.com/1405703/55539448-f378ca80-56c0-11e9-9b54-e43ebd1bda70.jpg)

kittaakos commented:
[minor]: Is there a chance to use the same binding for `Save As...`?

VS Code:
![Screen Shot 2019-04-04 at 10 04 59](https://user-images.githubusercontent.com/1405703/55539583-48b4dc00-56c1-11e9-8f20-b0fb36987c2b.jpg)

Theia:
![Screen Shot 2019-04-04 at 10 05 14](https://user-images.githubusercontent.com/1405703/55539572-43f02800-56c1-11e9-9902-f16b2bcd10d7.jpg)

kittaakos commented:
It feels a little bit unreliable. I have `auto-save` off, a dirty editor and I still cannot see the `Save As...` item in the `File` menu.

![Screen Shot 2019-04-04 at 10 08 58](https://user-images.githubusercontent.com/1405703/55539787-ce388c00-56c1-11e9-8e25-d2065affc8b0.jpg)

kittaakos commented:
> VS Code tells me the following:

Now I get the native do you want to override dialog, I pressed yes, I got another do you want to override dialog, not the native one, I pressed yes and nothing happened.

![Screen Shot 2019-04-04 at 10 10 54](https://user-images.githubusercontent.com/1405703/55539912-18217200-56c2-11e9-9535-fd8deea428a6.jpg)

```
/Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:490 Uncaught (in promise) TypeError: Cannot read property 'toString' of undefined
    at ElectronFileDialogService.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:490:73)
    at step (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:362:23)
    at Object.next (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:343:53)
    at file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:337:71
    at new Promise (<anonymous>)
    at push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.__awaiter (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:333:12)
    at ElectronFileDialogService.push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.ElectronFileDialogService.canReadWrite (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:479:16)
    at file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:464:51
    at CallbacksRegistry.apply (/Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/common/api/callbacks-registry.js:47:25)
    at EventEmitter.ipcRenderer.on (/Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:277:21)
```

vince-fugnitto commented:
> I cannot see `Save As...` in the `File` menu. It is visible from the Command Palette. Is it a bug or a feature?

There is a general problem with Electron menus, a lot of the menu items are not present until an update is made on them. For instance, here is the browser menu:

![Screenshot from 2019-04-04 06-45-16](https://user-images.githubusercontent.com/40359487/55550256-c539c180-56a5-11e9-84f0-4bb2540457a1.png)

Compared to Electron, the following are missing on startup:
1. New File
2. New Folder
3. Upload Files...
5. Download

vince-fugnitto commented:
> [minor]: Is there a chance to use the same binding for `Save As...`?

Sure, I had <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>s</kbd> since VSCode on Linux used it but I'll align for Mac.

vince-fugnitto commented:
> Now I get the native do you want to override dialog, I pressed yes, I got another do you want to override dialog, not the native one, I pressed yes and nothing happened.

The same occurs for the command `Save Workspace As...` currently in master, when fixing it I'll fix both.

vince-fugnitto commented:
> > [minor]: Is there a chance to use the same binding for `Save As...`?
> 
> Sure, I had ctrl+shift+s since VSCode on Linux used it but I'll align for Mac as well.

@kittaakos the keybinding for mac is already used by the `Toggle Outline View`. 
Do I align to the `Save As...` instead?

kittaakos commented:
> Do I align to the `Save As...` instead?

No, please leave as is. Thanks for checking it!

kittaakos commented:
I have the same error when I try to `Save As...` to an existing file:

```
/Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:490 Uncaught (in promise) TypeError: Cannot read property 'toString' of undefined
    at ElectronFileDialogService.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:490:73)
    at step (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:362:23)
    at Object.next (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:343:53)
    at file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:337:71
    at new Promise (<anonymous>)
    at push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.__awaiter (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:333:12)
    at ElectronFileDialogService.push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.ElectronFileDialogService.canReadWrite (file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:479:16)
    at file:///Users/akos.kitta/git/theia/examples/electron/lib/46.bundle.js:464:51
    at CallbacksRegistry.apply (/Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/common/api/callbacks-registry.js:47:25)
    at EventEmitter.ipcRenderer.on (/Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:277:21)
```

vince-fugnitto commented:
@kittaakos can you re-test for me whenever you get a chance?
I didn't find any errors when re-testing with the latest changes off master.

kittaakos commented:
I am verifying this once more. Sorry for the delay...

kittaakos commented:
It still does not work for me locally, I did the following:
 - build the electron application from your branch (5134375c2) locally.
 - open Theia.
 - create a new empty folder for the new temp Theia workspace (I used `mkdir`)
 - create a new file > `Untitled.txt`.
 - edit the file (auto-save if off), and set the selection here and there to have the`Save As...` file menu item.
 - save `Untitled.txt` as `Untitled.txt`.
 - I got the (native) replace existing dialog, I hit replace.

```
For more information and help, consult https://electronjs.org/docs/tutorial/security.
This warning will not show up once the app is packaged.
warnAboutInsecureCSP @ /Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/security-warnings.js:188
bundle.js:93642 root WARN LanguagesFrontendContribution.onStart is slow, took: 107.50999999999976 ms
log @ bundle.js:93642
bundle.js:93642 root WARN Preference minds.integration.atlas in user_storage:settings.json is invalid.
log @ bundle.js:93642
bundle.js:93642 root WARN Preference minds.integration.cmodel.download in user_storage:settings.json is invalid.
log @ bundle.js:93642
bundle.js:93642 root WARN Preference minds.integration.cmodel.upload in user_storage:settings.json is invalid.
log @ bundle.js:93642
bundle.js:93642 root WARN Preference minds.diagram.editor.lockCodeEditor in user_storage:settings.json is invalid.
log @ bundle.js:93642
bundle.js:93642 root WARN EditorNavigationContribution.onStart is slow, took: 425.58500000000004 ms
log @ bundle.js:93642
bundle.js:93642 root INFO Config file tasks.json does not exist under file:///Users/akos.kitta/Desktop/foo
bundle.js:93642 root WARN Couldn't restore widget for code-editor-opener. Error: Error: 'file:///Users/akos.kitta/Desktop/foo/Untitled.txt' has not been found. 
log @ bundle.js:93642
logger-protocol.ts:112 root ERROR onWillSave listeners should provide edits, not directly alter the document.
log @ logger-protocol.ts:112
(anonymous) @ logger-frontend-module.ts:41
(anonymous) @ logger.ts:312
(anonymous) @ logger.ts:304
Promise resolved (async)
../../packages/core/lib/common/logger.js.Logger.log @ logger.ts:299
log @ logger.ts:45
(anonymous) @ monaco-editor-model.js:486
step @ monaco-editor-model.js:58
(anonymous) @ monaco-editor-model.js:39
fulfilled @ monaco-editor-model.js:30
Promise resolved (async)
step @ monaco-editor-model.js:32
(anonymous) @ monaco-editor-model.js:33
push.../../packages/monaco/lib/browser/monaco-editor-model.js.__awaiter @ monaco-editor-model.js:29
(anonymous) @ monaco-editor-model.js:446
(anonymous) @ event.ts:231
step @ event.ts:15
(anonymous) @ event.ts:15
fulfilled @ event.ts:15
Promise resolved (async)
step @ event.ts:15
fulfilled @ event.ts:15
Promise resolved (async)
step @ event.ts:15
(anonymous) @ event.ts:15
../../packages/core/lib/common/event.js.__awaiter @ event.ts:15
../../packages/core/lib/common/event.js.Emitter.sequence @ event.ts:228
(anonymous) @ monaco-editor-model.js:446
step @ monaco-editor-model.js:58
(anonymous) @ monaco-editor-model.js:39
(anonymous) @ monaco-editor-model.js:33
push.../../packages/monaco/lib/browser/monaco-editor-model.js.__awaiter @ monaco-editor-model.js:29
push.../../packages/monaco/lib/browser/monaco-editor-model.js.MonacoEditorModel.fireWillSaveModel @ monaco-editor-model.js:440
(anonymous) @ monaco-editor-model.js:415
step @ monaco-editor-model.js:58
(anonymous) @ monaco-editor-model.js:39
(anonymous) @ monaco-editor-model.js:33
push.../../packages/monaco/lib/browser/monaco-editor-model.js.__awaiter @ monaco-editor-model.js:29
push.../../packages/monaco/lib/browser/monaco-editor-model.js.MonacoEditorModel.doSave @ monaco-editor-model.js:407
(anonymous) @ monaco-editor-model.js:361
(anonymous) @ monaco-editor-model.js:255
step @ monaco-editor-model.js:58
(anonymous) @ monaco-editor-model.js:39
(anonymous) @ monaco-editor-model.js:33
push.../../packages/monaco/lib/browser/monaco-editor-model.js.__awaiter @ monaco-editor-model.js:29
(anonymous) @ monaco-editor-model.js:249
Promise resolved (async)
push.../../packages/monaco/lib/browser/monaco-editor-model.js.MonacoEditorModel.run @ monaco-editor-model.js:249
push.../../packages/monaco/lib/browser/monaco-editor-model.js.MonacoEditorModel.scheduleSave @ monaco-editor-model.js:361
push.../../packages/monaco/lib/browser/monaco-editor-model.js.MonacoEditorModel.save @ monaco-editor-model.js:245
(anonymous) @ saveable.ts:69
step @ saveable.ts:15
(anonymous) @ saveable.ts:15
(anonymous) @ saveable.ts:15
../../packages/core/lib/browser/saveable.js.__awaiter @ saveable.ts:15
save @ saveable.ts:66
(anonymous) @ application-shell.ts:1384
step @ application-shell.ts:15
(anonymous) @ application-shell.ts:15
(anonymous) @ application-shell.ts:15
../../packages/core/lib/browser/shell/application-shell.js.__awaiter @ application-shell.ts:15
../../packages/core/lib/browser/shell/application-shell.js.ApplicationShell.save @ application-shell.ts:1383
execute @ common-frontend-contribution.ts:445
../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.tryKeybindingExecution @ keybinding.ts:486
../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.run @ keybinding.ts:534
(anonymous) @ frontend-application.ts:171
electron-file-dialog-service.ts:86 Uncaught (in promise) TypeError: Cannot read property 'toString' of undefined
    at ElectronFileDialogService.<anonymous> (electron-file-dialog-service.ts:86)
    at step (electron-file-dialog-service.ts:15)
    at Object.next (electron-file-dialog-service.ts:15)
    at electron-file-dialog-service.ts:15
    at new Promise (<anonymous>)
    at push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.__awaiter (electron-file-dialog-service.ts:15)
    at ElectronFileDialogService.push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.ElectronFileDialogService.canReadWrite (electron-file-dialog-service.ts:84)
    at electron-file-dialog-service.ts:73
    at CallbacksRegistry.apply (/Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/common/api/callbacks-registry.js:47)
    at EventEmitter.ipcRenderer.on (/Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:277)
(anonymous) @ electron-file-dialog-service.ts:86
step @ electron-file-dialog-service.ts:15
(anonymous) @ electron-file-dialog-service.ts:15
(anonymous) @ electron-file-dialog-service.ts:15
push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.__awaiter @ electron-file-dialog-service.ts:15
push.../../packages/filesystem/lib/electron-browser/file-dialog/electron-file-dialog-service.js.ElectronFileDialogService.canReadWrite @ electron-file-dialog-service.ts:84
(anonymous) @ electron-file-dialog-service.ts:73
apply @ /Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/common/api/callbacks-registry.js:47
ipcRenderer.on @ /Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:277
emitMany @ events.js:147
emit @ events.js:224
messages.js:46 Uncaught (in promise) Error: Request copy failed with message: Source and destination must not be the same.
    at Proxy.<anonymous> (proxy-factory.ts:232)
    at WorkspaceFrontendContribution.<anonymous> (workspace-frontend-contribution.ts:360)
    at step (workspace-frontend-contribution.ts:15)
    at Object.next (workspace-frontend-contribution.ts:15)
    at fulfilled (workspace-frontend-contribution.ts:15)
    at <anonymous>
ResponseError @ messages.js:46
handleResponse @ main.js:430
processMessageQueue @ main.js:258
(anonymous) @ main.js:242
runCallback @ timers.js:789
tryOnImmediate @ timers.js:751
processImmediate @ timers.js:722
Promise rejected (async)
(anonymous) @ proxy-factory.ts:242
Promise rejected (async)
(anonymous) @ proxy-factory.ts:242
(anonymous) @ proxy-factory.ts:234
Async Call
(anonymous) @ proxy-factory.ts:233
(anonymous) @ workspace-frontend-contribution.ts:360
step @ workspace-frontend-contribution.ts:15
(anonymous) @ workspace-frontend-contribution.ts:15
fulfilled @ workspace-frontend-contribution.ts:15
Promise resolved (async)
step @ workspace-frontend-contribution.ts:15
fulfilled @ workspace-frontend-contribution.ts:15
Promise resolved (async)
step @ workspace-frontend-contribution.ts:15
fulfilled @ workspace-frontend-contribution.ts:15
Promise resolved (async)
step @ workspace-frontend-contribution.ts:15
(anonymous) @ workspace-frontend-contribution.ts:15
push.../../packages/workspace/lib/browser/workspace-frontend-contribution.js.__awaiter @ workspace-frontend-contribution.ts:15
push.../../packages/workspace/lib/browser/workspace-frontend-contribution.js.WorkspaceFrontendContribution.saveAs @ workspace-frontend-contribution.ts:338
execute @ workspace-frontend-contribution.ts:94
../../packages/core/lib/common/uri-command-handler.js.UriAwareCommandHandler.execute @ uri-command-handler.ts:89
(anonymous) @ command.ts:241
step @ command.ts:15
(anonymous) @ command.ts:15
(anonymous) @ command.ts:15
../../packages/core/lib/common/command.js.__awaiter @ command.ts:15
../../packages/core/lib/common/command.js.CommandRegistry.executeCommand @ command.ts:238
(anonymous) @ electron-main-menu-factory.ts:220
step @ electron-main-menu-factory.ts:15
(anonymous) @ electron-main-menu-factory.ts:15
(anonymous) @ electron-main-menu-factory.ts:15
push.../../packages/core/lib/electron-browser/menu/electron-main-menu-factory.js.__awaiter @ electron-main-menu-factory.ts:15
push.../../packages/core/lib/electron-browser/menu/electron-main-menu-factory.js.ElectronMainMenuFactory.execute @ electron-main-menu-factory.ts:213
click @ electron-main-menu-factory.ts:140
apply @ /Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/common/api/callbacks-registry.js:47
ipcRenderer.on @ /Users/akos.kitta/git/theia/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:277
emitMany @ events.js:147
emit @ events.js:224
```

I can see this popping up, I do not know why does the override do a copy under the hood:
```
root ERROR Request copy failed with error: Source and destination must not be the same. Error: Source and destination must not be the same.
    at Object.copy (/Users/akos.kitta/git/theia/node_modules/fs-extra/lib/copy/copy.js:29:51)
    at Promise (/Users/akos.kitta/git/theia/node_modules/universalify/index.js:13:12)
    at new Promise (<anonymous>)
    at Object.copy (/Users/akos.kitta/git/theia/node_modules/universalify/index.js:7:14)
    at FileSystemNode.<anonymous> (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:413:49)
    at step (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:59:23)
    at Object.next (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:40:53)
    at fulfilled (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:31:58)
    at <anonymous>
```

vince-fugnitto commented:
@kittaakos sorry I added the handling code and rebased off master, I guess my force push before didn't work properly. I followed your example and everything worked fine on my end.

I also rebuilt using the node 10, and performed `yarn rebuild:electron` before starting the electron application.

kittaakos commented:
I am trying it...

vince-fugnitto commented:
@kittaakos just a question, are you testing on windows?
I've tried both on linux and mac and I can't seem to reproduce the error :disappointed: 

The error you referenced has to do with checking if you can `canReadWrite` to that uri, and 
somehow the uri is `undefined`.

vince-fugnitto commented:
@kittaakos sorry for bothering you with the PR over the past week.
I have found the reason why it was misbehaving, I wasn't properly performing an `await` for the `copy`.

The error is now caught and logged, but I was wondering, is it sufficient to simply log the error message, or should we keep the stack trace as well?

vince-fugnitto commented:
> ```diff
> diff --git a/packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-service.ts b/packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-service.ts
> index fabe0cd77..6aa3985ed 100644
> --- a/packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-service.ts
> +++ b/packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-service.ts
> @@ -82,7 +82,7 @@ export class ElectronFileDialogService extends DefaultFileDialogService {
>      }
>  
>      protected async canReadWrite(uris: MaybeArray<URI>): Promise<boolean> {
> -        for (const uri of uris) {
> +        for (const uri of Array.isArray(uris) ? uris : [uris]) {
>              if (!(await this.fileSystem.access(uri.toString(), FileAccess.Constants.R_OK | FileAccess.Constants.W_OK))) {
>                  this.messageService.error(`Cannot access resource at ${uri.path}.`);
>                  return false;
> ```


Thank you! I updated the code based on your proposed changes.

akosyakov commented on packages/workspace/src/browser/workspace-frontend-contribution.ts:
```diff
@@ -314,6 +330,41 @@ export class WorkspaceFrontendContribution implements CommandContribution, Keybi
+                if (exist) {
+                    // Prompt users for confirmation before overwriting.
+                    // Electron already handles the confirmation so do not prompt again.
+                    overwrite = this.isElectron() ? true : await this.confirmOverwrite(target);
```
fyi: you could move electron check in `confirmOverwrite` to avoid duplication

kittaakos commented:
No errors this time, but it does not work.
![screencast 2019-04-10 09-17-11](https://user-images.githubusercontent.com/1405703/55859119-accc1a00-5b71-11e9-9e7e-2cb17bffd4b2.gif)

vince-fugnitto commented:
> No errors this time, but it does not work.
> ![screencast 2019-04-10 09-17-11](https://user-images.githubusercontent.com/1405703/55859119-accc1a00-5b71-11e9-9e7e-2cb17bffd4b2.gif)

I found it a bit odd the first time I checked, but in VSCode the behavior is the same.
They do not apply the dirty changes when performing `Save As...` and actually rollback to the clean state. (both when making a copy or when overwriting)

vince-fugnitto commented:
@kittaakos are there any more requested updates for the pr?

kittaakos commented:
> No errors this time, but it does not work.

Was this fixed?

As a user, I would expect when I do `Save As...` and I select the same file, my editor will be saved, and the corresponding file change event will be triggered. Otherwise, my LS won't get the file change event.

vince-fugnitto commented:
> Was this fixed?

It should be resolved in the latest changes :smiley:

kittaakos commented:
> It should be resolved in the latest changes 😃

Great, thanks! I'll verify it later today.

kittaakos commented:
It works as expected 👍 Can we log an `info/warning` if we `Save as...` to the same file? Now it is,
```
root ERROR Error: Request 'copy' failed
    at Proxy.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:95932:33)
    at WorkspaceFrontendContribution.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/8.bundle.js:2544:62)
    at step (file:///Users/akos.kitta/git/theia/examples/electron/lib/8.bundle.js:2098:23)
    at Object.next (file:///Users/akos.kitta/git/theia/examples/electron/lib/8.bundle.js:2079:53)
    at fulfilled (file:///Users/akos.kitta/git/theia/examples/electron/lib/8.bundle.js:2070:58)
Caused by: Error: Cannot perform copy, source and destination are the same. 'file:///Users/akos.kitta/Desktop/tmp-save-as/Untitled.txt' already exists.
    at FileSystemNode.<anonymous> (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:414:64)
    at step (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:59:23)
    at Object.next (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:40:53)
    at fulfilled (/Users/akos.kitta/git/theia/packages/filesystem/lib/node/node-filesystem.js:31:58)
```

vince-fugnitto commented:
> It works as expected 👍 Can we log an `info/warning` if we `Save as...` to the same file? Now it is,

Sure! 😄

vince-fugnitto commented:
@kittaakos just wanted to confirm everything was okay, and that it'd be fine to merge :)

kittaakos commented:
> it'd be fine to merge

👍
